### PR TITLE
Alias Delete Objects Via Config

### PR DESCRIPTION
**Pull request by Phillita**
# WHAT
Deletes fail when SFDC sends for an object to delete that has a different name in the app. IE. `Account` has been aliased to `Vendor`.

# HOW
Added a salesforce_ar_sync.yml option to allow you to specify a map of SFDC objects to app objects.

# OTHER
Made a lot of Rubocop changes to the gem... I just had to!

### CODE REVIEW

File: lib/salesforce_ar_sync/engine.rb [hunk 1]
```diff
@@ -1,29 +1,31 @@
...
           config = YAML.load_file("#{Rails.root}/config/salesforce_ar_sync.yml")[Rails.env]
-          SalesforceArSync.config["ORGANIZATION_ID"] = config['organization_id']
-          SalesforceArSync.config["SYNC_ENABLED"] = config['sync_enabled']
-          SalesforceArSync.config["IP_RANGES"] = config['ip_ranges'].split(',').map{ |ip| ip.strip }
-          SalesforceArSync.config["NAMESPACE_PREFIX"] = config['namespace_prefix']
+          SalesforceArSync.config['ORGANIZATION_ID'] = config['organization_id']
+          SalesforceArSync.config['SYNC_ENABLED'] = config['sync_enabled']
+          SalesforceArSync.config['IP_RANGES'] = config['ip_ranges'].split(',').map(&:strip)
+          SalesforceArSync.config['NAMESPACE_PREFIX'] = config['namespace_prefix']
+          SalesforceArSync.config['DELETION_MAP'] = config['deletion_map'].stringify_keys
```
Comment (Phillita) [inline lib/salesforce_ar_sync/engine.rb]: Changed this.

File: lib/salesforce_ar_sync/soap_handler/delete.rb [hunk 2]
```diff
@@ -1,19 +1,21 @@
...
+      end
 
-       def self.delete_object(hash = {})
-         raise ArgumentError, "Object_Id__c parameter required" if hash[namespaced(:Object_Id__c)].blank?
-         raise ArgumentError, "Object_Type__c parameter required" if hash[namespaced(:Object_Type__c)].blank?
-     
-         object = hash[namespaced(:Object_Type__c)].constantize.find_by_salesforce_id(hash[namespaced(:Object_Id__c)])
-         object.destroy if object && object.ar_sync_inbound_delete?
-       end
+      def self.delete_object(hash = {})
```
Comment (Phillita) [inline lib/salesforce_ar_sync/soap_handler/delete.rb]: Changed this.

File: spec/salesforce_ar_sync/soap_handlers/delete_spec.rb [hunk 3]
```diff
@@ -0,0 +1,62 @@
+require 'spec_helper.rb'
```
Comment (Phillita) [inline spec/salesforce_ar_sync/soap_handlers/delete_spec.rb]: added specs for the delete.

**Comment by lnediger** [comment]
Is this necessary? The way it was designed was that the inbound delete message would indicate the name of the model that the delete is for.

File: lib/salesforce_ar_sync/extenders/salesforce_syncable.rb [hunk 4]
```diff
@@ -5,22 +5,22 @@ def salesforce_syncable(options = {})
...
-        self.salesforce_sync_attribute_mapping = options.has_key?(:sync_attributes) ? options[:sync_attributes].stringify_keys : {}
-        self.salesforce_async_attributes = options.has_key?(:async_attributes) ? options[:async_attributes] : {}
-        self.salesforce_default_attributes_for_create = options.has_key?(:default_attributes_for_create) ? options[:default_attributes_for_create] : {}
-        self.salesforce_id_attribute_name = options.has_key?(:salesforce_id_attribute_name) ? options[:salesforce_id_attribute_name] : :Id
-        self.salesforce_web_id_attribute_name = options.has_key?(:web_id_attribute_name) ? options[:web_id_attribute_name] : :WebId__c
-        self.activerecord_web_id_attribute_name = options.has_key?(:activerecord_web_id_attribute_name) ? options[:activerecord_web_id_attribute_name] : :id
-        self.salesforce_sync_web_id = options.has_key?(:salesforce_sync_web_id) ? options[:salesforce_sync_web_id] : false
-        self.salesforce_web_class_name = options.has_key?(:web_class_name) ? options[:web_class_name] : self.name
+        self.salesforce_sync_enabled = options.fetch(:salesforce_sync_enabled, true)
```
Comment (michaelbrawn) [inline lib/salesforce_ar_sync/extenders/salesforce_syncable.rb]: :+1:

File: spec/salesforce_ar_sync/salesforce_ar_sync_spec.rb [hunk 5]
```diff
@@ -22,30 +22,30 @@ def email_address_changed?
...
   # Hack for Parsing into the proper Timezone
   def salesforce_updated_at=(updated_at)
-    updated_at = Time.parse(updated_at) if updated_at.present? && updated_at.kind_of?(String)
+    updated_at = Time.parse(updated_at) if updated_at.present? && updated_at.is_a?(String)
     write_attribute(:salesforce_updated_at, updated_at)
   end
 end
 
 # the following hash should match the data in SF.com you are testing against
 sample_outbound_message_hash = {
```
Comment (michaelbrawn) [inline spec/salesforce_ar_sync/salesforce_ar_sync_spec.rb]: Would it make sense to change these hashes to constants?

File: spec/salesforce_ar_sync/soap_handlers/delete_spec.rb (hunk 3 above)
Comment (michaelbrawn) [inline spec/salesforce_ar_sync/soap_handlers/delete_spec.rb]: We should probably add a spec for when the DELETION_MAP isn't configured.

**Comment by michaelbrawn** [comment]
@lnediger This would allow us to configure the map in our app, instead of needing salesforce outbound messages to know which classes exist in our app. @Phillita may be able to speak to this more.

**Comment by Phillita** [comment]
@lnediger It sounds like we would have to make an outbound message in SFDC that knew about the Account to Vendor mapping. Currently, SFDC puts everything into one general deleted object table and then send them to the delete route. It just takes the type of object, in this case account, and send that to the delete without any sort of alias.

What are your thoughts?

**Comment by Phillita** [comment]
Changes: 
I'm going to pull the code changes out so that there isn't a mix of Syntax updates with it.
I'll also be updating the readme with the new config file updates.

Hopefully this will help with tests.

**Comment by Phillita** [comment]
Deleting this branch in favour of #19.

**Pull request closed without merging**